src/wallet/rpcdump.cpp

```diff
@@ -663,51 +662,40 @@ UniValue importwallet(const JSONRPCRequest& request)
             CPubKey pubkey = key.GetPubKey();
             assert(key.VerifyPubKey(pubkey));
             CKeyID keyid = pubkey.GetID();
-            if (pwallet->HaveKey(keyid)) {
-                pwallet->WalletLogPrintf("Skipping import of %s (key already present)\n", EncodeDestination(PKHash(keyid)));
+
+            std::map<CKeyID, CKey> privkey_map;
+            privkey_map.emplace(keyid, key);
+
+            if (!pwallet->ImportPrivKeys(privkey_map, time)) {
+                pwallet->WalletLogPrintf("Skipping import of %s (key already present or an error occurred)\n", EncodeDestination(PKHash(keyid)));
                 continue;
             }
+
             pwallet->WalletLogPrintf("Importing %s...\n", EncodeDestination(PKHash(keyid)));
-            if (!pwallet->AddKeyPubKey(key, pubkey)) {
-                fGood = false;
-                continue;
-            }
-            pwallet->mapKeyMetadata[keyid].nCreateTime = time;
-            if (has_label)
-                pwallet->SetAddressBook(PKHash(keyid), label, "receive");
+
+            pwallet->ImportScriptPubKeys(label, {GetScriptForDestination(PKHash(keyid))}, true /* have_solving_data */, has_label, time);
+
             nTimeBegin = std::min(nTimeBegin, time);
             progress++;
         }
         for (const auto& script_pair : scripts) {
             pwallet->chain().showProgress("", std::max(50, std::min(75, (int)((progress / total) * 100) + 50)), false);
             const CScript& script = script_pair.first;
             int64_t time = script_pair.second;
-            CScriptID id(script);
-            if (pwallet->HaveCScript(id)) {
-                pwallet->WalletLogPrintf("Skipping import of %s (script already present)\n", HexStr(script));
-                continue;
-            }
-            if(!pwallet->AddCScript(script)) {
-                pwallet->WalletLogPrintf("Error importing script %s\n", HexStr(script));
-                fGood = false;
+
+            if (pwallet->ImportScripts({script}, time)) {
```

Comment:
isn't this backwards?

Reply:
How so?

Reply:
Because it's reporting errors when it returns true?

Reply:
Oh, whoops. Fixed.